src/traces/heatmap/calc.js

```diff
@@ -0,0 +1,432 @@
+/**
+* Copyright 2012-2015, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+var isNumeric = require('fast-isnumeric');
+
+var Plotly = require('../../plotly');
+var Lib = require('../../lib');
+
+var histogram2dCalc = require('../histogram2d/calc');
+var hasColumns = require('./has_columns');
+var convertColumnXYZ = require('./convert_column_xyz');
+var maxRowLength = require('./max_row_length');
+
+
+module.exports = function calc(gd, trace) {
+    Lib.markTime('start convert x&y');
+
+    // prepare the raw data
+    // run makeCalcdata on x and y even for heatmaps, in case of category mappings
+    var xa = Plotly.Axes.getFromId(gd, trace.xaxis||'x'),
+        ya = Plotly.Axes.getFromId(gd, trace.yaxis||'y'),
+        isContour = Plotly.Plots.traceIs(trace, 'contour'),
+        isHist = Plotly.Plots.traceIs(trace, 'histogram'),
+        zsmooth = isContour ? 'best' : trace.zsmooth,
+        x,
+        x0,
+        dx,
+        y,
+        y0,
+        dy,
+        z,
+        i;
+
+    // cancel minimum tick spacings (only applies to bars and boxes)
+    xa._minDtick = 0;
+    ya._minDtick = 0;
+
+    Lib.markTime('done convert x&y');
+
+    if(isHist) {
+        var binned = histogram2dCalc(gd, trace);
```

Comment:
To keep this PR readable, histogram2d and heatmap still share the same calc step.